packages/flame/lib/src/components/sprite_group_component.dart

```diff
@@ -100,7 +100,7 @@ class SpriteGroupComponent<T> extends PositionComponent
 
   /// Updates the sprite for the given key.
   void updateSprite(T key, Sprite sprite) {
-    _sprites![key] = sprite;
+    _sprites?[key] = sprite;
```

Comment:
We should probably have an assertion here that checks that the map is initialized.

Reply:
Actually, the map ```_sprites``` initialize in the ```onMount()``` method of the ```SpriteButtonComponent ``` class.

If we use the constructor ```SpriteButtonComponent()```, ```updateSprite(T key, Sprite sprite)``` method isn't used.
and if we set the ```button``` setter in the ```onLoad()``` method of the ```SpriteButtonComponet``` class, ```updateSprite(T key, Sprite sprite)``` method was called before the ```onMount()```,

So before ```_sprites``` initialization, ```_sprites![key] = sprite;```, line code was executing.
So I think assertion is not required here, as assertion is there for ```button``` and ```sprites``` of ``` SpriteButtonComponent``` and ```SpriteGroupComponent``` class in their respective ```onMount()``` method.

Reply:
You could call `updateSprite` before `onLoad` (when you're initializing the map in there) and then it would just be skipped and the user wouldn't know, hence why the assertion is needed, so that the user can be informed that they made an unintentional no-op.

Reply:
I didn't get the point how ```updateSprite``` can be called before ```onLoad``` method. 
because when we use the ```SpriteButtonComponent``` in a project and set the ```Sprite``` using ```button``` setter (in ```onLoad()``` method of our class), then the ```updateSprite``` method called. 

` set button(Sprite value) {`
  `  _button = value;`
    `updateSprite(ButtonState.up, value);`
 ` }`

and if we add assertion ```_sprites != null```, in ```updateSprite``` method, it will throw error every time, no matter ```Sprite``` is set or not (because map initialize in ```onMount()```). (tested).

So by replacing ```!```(null check) with ```?``` (null aware) operator, it will not throw error and ```_sprites``` are going to initialize in the ```onMount()``` method of the ```SpriteButtonComponent``` and assertion is doing their job in the same method. (it is working fine).

Am I missing something. : )

Reply:
Since `updateSprite` isn't a private or `internal` method, it can be called directly by the user too, the `SpriteGroupComponent` has many more use-cases than the `SpriteButtonComponent`.
You have discovered a bug in the `SpriteButtonComponent` though, the `sprites` map should be initialized in `onLoad`, just like it says in the `onMount` assertion in `SpriteGroupComponent`.
So I'd suggest you add an assertion in `updateSprite` and move the initialization of the `sprites` map from `onMount` to `onLoad` in `SpriteButtonComponent`.

Reply:
I fixed it up in [26514a1](https://github.com/flame-engine/flame/pull/3302/commits/26514a16bf79d4186e65cb236631374fa503f62b)

Reply:
Changes made in [26514a1](https://github.com/flame-engine/flame/pull/3302/commits/26514a16bf79d4186e65cb236631374fa503f62b) actually throws assertion `Sprites map is not yet initialized` every time when we use the `SpriteButtonComponent` class. 
Moving `sprites` map initialization from `onMount` to `onLoad` doesn't make much difference. If we consider an example 

```
class DotButtonComponent extends SpriteButtonComponent with HasGameReference<Pitu> {
  @override
  Future<void> onLoad() async {
    // super.onLoad();
    position = game.size / 2;
    anchor = Anchor.center;
    // autoResize = true;
    button = Sprite(Flame.images.fromCache('dot.png'));
    buttonDown = Sprite(Flame.images.fromCache('odd.jpg'));
    onPressed = () {
      log("Pressed");
    };
    super.onLoad();
  }
}
```
here the `button` setter will execute first (before `super.onLoad()`) and so the `updateSprite` method (will throw assertion),
So the user will not know the cause of the assertion as `Sprites` have already been set with `button` and `buttonDown` setter.

Reply:
`super.onLoad` should always be called first in all Flame components, and since the assertion says that it should be initialized in `onLoad` it should be pretty clear to the user that they are running code before the parent's `onLoad` has run. If you have any better wording for the assertion, feel free to open a PR with it.

Reply:
Ah, I thought I mentioned `onLoad` + the constructor in the assertion, apparently I only did that in the other assertion.

Reply:
Clarified it even more in here:
https://github.com/flame-engine/flame/pull/3317

Reply:
Actually, even if we call the `super.onLoad` before the rest of the code (especially, `button` setter), we got null check error in button setter,
```
class DotButtonComponent extends SpriteButtonComponent with HasGameReference<Pitu> {
  @override
  Future<void> onLoad() async {
    super.onLoad();
    position = game.size / 2;
    anchor = Anchor.center;
    // autoResize = true;
    button = Sprite(Flame.images.fromCache('dot.png'));
    buttonDown = Sprite(Flame.images.fromCache('odd.jpg'));
    onPressed = () {
      log("Pressed");
    };
  }
}
```
As the `onLoad` will execute first. (`SpriteButtonComponent`)
```
 @override
  void onLoad() {
    super.onLoad();
    sprites = {
      ButtonState.up: button,
      ButtonState.down: buttonDown,
    };
  }
``` 
here the `button` is not yet set. so `Sprite get button => _button!;` will throw error because `_button` is null but `!`(null check) is used here.